java/adapter/avro/src/main/java/org/apache/arrow/AvroToArrowUtils.java

```diff
@@ -75,29 +76,61 @@
    *   <li>BYTES --> ArrowType.Binary</li>
    * </ul>
    */
-  private static ArrowType getArrowType(Type type) {
+  private static Field getArrowField(Schema schema, String name, boolean nullable) {
 
-    Preconditions.checkNotNull(type, "Avro type object can't be null");
+    Preconditions.checkNotNull(schema, "Avro schema object can't be null");
+
+    Type type = schema.getType();
+    ArrowType arrowType;
 
     switch (type) {
+      case UNION:
+        return getUnionField(schema, name);
       case STRING:
-        return new ArrowType.Utf8();
+        arrowType = new ArrowType.Utf8();
+        break;
       case INT:
-        return new ArrowType.Int(32, /*signed=*/true);
+        arrowType = new ArrowType.Int(32, /*signed=*/true);
+        break;
       case BOOLEAN:
-        return new ArrowType.Bool();
+        arrowType = new ArrowType.Bool();
+        break;
       case LONG:
-        return new ArrowType.Int(64, /*signed=*/true);
+        arrowType = new ArrowType.Int(64, /*signed=*/true);
+        break;
       case FLOAT:
-        return new ArrowType.FloatingPoint(SINGLE);
+        arrowType =  new ArrowType.FloatingPoint(SINGLE);
+        break;
       case DOUBLE:
-        return new ArrowType.FloatingPoint(DOUBLE);
+        arrowType = new ArrowType.FloatingPoint(DOUBLE);
+        break;
       case BYTES:
-        return new ArrowType.Binary();
+        arrowType = new ArrowType.Binary();
+        break;
       default:
         // no-op, shouldn't get here
         throw new RuntimeException("Can't convert avro type %s to arrow type." + type.getName());
     }
+
+    final FieldType fieldType =  new FieldType(/*nullable=*/nullable, arrowType, /*dictionary=*/null,
+        /*metadata=*/getMetaData(schema));
+    return new Field(name, fieldType, null);
+  }
+
+  private static Field getUnionField(Schema schema, String name) {
+    int size = schema.getTypes().size();
+    long nullCount = schema.getTypes().stream().filter(s -> s.getType() == Type.NULL).count();
+    // if has two field and one is null type, convert to nullable primitive type
+    if (size == 2 && nullCount == 1) {
```

Comment:
can types repeat in Avro?  Should this be size == nullCount + 1

Reply:
Repeated types are not allowed. Right, size == nullCount + 1 could check nullable and size == 2 could check nullable primitive type, fixed.